File: lld/concurrency/java/src/ConcurrencyTest.java
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Plain assert-based test runner (no JUnit dependency needed).
 * Run via `java -cp out Main` (invoked from Main) or `java -cp out ConcurrencyTest` directly.
 */
public class ConcurrencyTest {

    public static void main(String[] args) {
        runAll();
    }

    public static void runAll() {
        testNoLostOrDuplicatedItems();
        testSingleProducerSingleWorker();
        System.out.println("All ConcurrencyTest cases passed.");
    }

    private static void testNoLostOrDuplicatedItems() {
        final int numProducers = 5;
        final int itemsPerProducer = 200;
        final int numWorkers = 8;
        final int bufferCapacity = 16;

        ConcurrentHashMap<String, Integer> counts = new ConcurrentHashMap<>();
        BoundedPipeline pipeline = new BoundedPipeline(numWorkers, bufferCapacity);
        pipeline.run(numProducers, itemsPerProducer, item ->
                counts.merge(item.id(), 1, Integer::sum));

        int wantTotal = numProducers * itemsPerProducer;
        assertEquals(wantTotal, counts.size(), "unique items consumed");

        for (int p = 0; p < numProducers; p++) {
            for (int seq = 0; seq < itemsPerProducer; seq++) {
                String id = "p" + p + "-" + seq;
                Integer count = counts.get(id);
                if (count == null) {
                    throw new AssertionError("item " + id + " was never consumed");
                }
                assertEquals(1, count, "delivery count for " + id);
            }
        }
    }

    private static void testSingleProducerSingleWorker() {
        Set<String> consumed = ConcurrentHashMap.newKeySet();
        BoundedPipeline pipeline = new BoundedPipeline(1, 1);
        pipeline.run(1, 50, item -> consumed.add(item.id()));
        assertEquals(50, consumed.size(), "items consumed with single producer/worker");
    }

    private static void assertEquals(Object expected, Object actual, String label) {
        if (!expected.equals(actual)) {
            throw new AssertionError(label + ": expected " + expected + " but got " + actual);
        }
    }
}

File: lld/concurrency/go/concurrency_test.go
package concurrency

import (
	"fmt"
	"testing"
)

// TestNoLostOrDuplicatedItems drives multiple producers and multiple worker
// consumers through a bounded channel and asserts every produced item is
// consumed exactly once: no loss, no duplication, regardless of scheduling.
func TestNoLostOrDuplicatedItems(t *testing.T) {
	const numProducers = 5
	const itemsPerProducer = 200
	const numWorkers = 8
	const bufferSize = 16

	results := NewSafeSet()
	pool := &Pool{NumWorkers: numWorkers, BufferSize: bufferSize}
	pool.Run(numProducers, itemsPerProducer, func(item Item) {
		results.Add(item.ID())
	})

	snapshot := results.Snapshot()
	wantTotal := numProducers * itemsPerProducer
	if len(snapshot) != wantTotal {
		t.Fatalf("expected %d unique items consumed, got %d", wantTotal, len(snapshot))
	}

	for pid := 0; pid < numProducers; pid++ {
		for seq := 0; seq < itemsPerProducer; seq++ {
			id := fmt.Sprintf("p%d-%d", pid, seq)
			count, ok := snapshot[id]
			if !ok {
				t.Fatalf("item %s was never consumed", id)
			}
			if count != 1 {
				t.Fatalf("item %s consumed %d times, want exactly 1", id, count)
			}
		}
	}
}

// TestSingleProducerSingleWorker sanity-checks the pipeline degrades
// correctly to the simplest 1:1 case.
func TestSingleProducerSingleWorker(t *testing.T) {
	results := NewSafeSet()
	pool := &Pool{NumWorkers: 1, BufferSize: 1}
	pool.Run(1, 50, func(item Item) {
		results.Add(item.ID())
	})

	if got := len(results.Snapshot()); got != 50 {
		t.Fatalf("expected 50 items consumed, got %d", got)
	}
}
